// eslint-plugin-jsdoc/dist/rules/noDefaults.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _iterateJsdoc = _interopRequireDefault(require("../iterateJsdoc"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = (0, _iterateJsdoc.default)(({
  context,
  utils
}) => {
  const _ref = context.options[0] || {},
        noOptionalParamNames = _ref.noOptionalParamNames;

  const paramTags = utils.getPresentTags(['param', 'arg', 'argument']);
  paramTags.forEach(tag => {
    if (noOptionalParamNames && tag.optional) {
      utils.reportJSDoc(`Optional param names are not permitted on @${tag.tag}.`, tag, () => {
        tag.default = '';
        tag.optional = false;
      });
    } else if (tag.default) {
      utils.reportJSDoc(`Defaults are not permitted on @${tag.tag}.`, tag, () => {
        tag.default = '';
      });
    }
  });
  const defaultTags = utils.getPresentTags(['default', 'defaultvalue']);
  defaultTags.forEach(tag => {
    if (tag.description) {
      utils.reportJSDoc(`Default values are not permitted on @${tag.tag}.`, tag, () => {
        tag.description = '';
      });
    }
  });
}, {
  contextDefaults: true,
  meta: {
    fixable: 'code',
    schema: [{
      additionalProperties: false,
      properties: {
        contexts: {
          items: {
            type: 'string'
          },
          type: 'array'
        },
        noOptionalParamNames: {
          type: 'boolean'
        }
      },
      type: 'object'
    }],
    type: 'suggestion'
  }
});

exports.default = _default;
module.exports = exports.default;
